tools/generator/filters/haskell.py

```diff
@@ -0,0 +1,36 @@
+from functools import partial
+
+from . import register_filter
+from .common import generic_comment, get_array_inner, is_array
+
+
+@register_filter
+def haskell_enum_fields(fields) -> str:
+    return '\n  | '.join(f'{field[0].capitalize()} -- {field[1]}'
+            for field in fields)
+
+
+@register_filter
+def haskell_type(type_id: str) -> str:
+    if is_array(type_id):
+        return f'[{haskell_type(get_array_inner(type_id))}]'
+    elif type_id == 'void':
+        return '()'
+    return type_id.capitalize()
+
+
+def _c_haskell_type(type_id: str) -> str:
```

Comment:
You can put this function inside `c_haskell_type` with a name like `recurse_inner` or something like that, since it's not meant to be used by anything else than the function itself.